### PR TITLE
ARROW-15721: [Docs][FlightRPC] Add Flight/Flight SQL to subprojects

### PR DESCRIPTION
**Pull request by lidavidm**
Adds a Flight SQL documentation page that we can link to from the website.

### CODE REVIEW

**Comment by lidavidm** [comment]
I would like to add sequence diagrams, how do we feel about something like [sphinxcontrib-mermaid](https://pypi.org/project/sphinxcontrib-mermaid/)?

**Comment by lidavidm** [comment]
CC @jduo if you'd like to take a look

File: docs/source/format/FlightSql.rst [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+===========
+
+Flight SQL reuses the predefined RPC methods in Arrow Flight, and
+provides various commands that pair those methods with request/response
+messages defined via Protobuf (see below).
+
+SQL Metadata
+------------
+
+Flight SQL provides a variety of commands to fetch metadata about the
```
Comment (jduo) [inline docs/source/format/FlightSql.rst:46]:
```suggestion
Flight SQL provides a variety of commands to fetch catalog metadata about the
```

File: docs/source/format/FlightSql.rst [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+``CommandGetCatalogs``
+    List the catalogs available in the database. The definition varies
+    by vendor.
+
+``CommandGetCrossReference``
+    List the foreign key columns in a given table that reference
+    columns in a given parent table.
+
+``CommandGetDbSchemas``
+    List the schemas (note: *not* an Arrow schema) available in the
```
Comment (jduo) [inline docs/source/format/FlightSql.rst:72]:
```suggestion
    List the schemas (note: a grouping of tables, *not* an Arrow schema) available in the
```

File: docs/source/format/FlightSql.rst [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+================
+
+Arrow Flight SQL is a protocol for interacting with SQL databases
+using the Arrow in-memory format and the :doc:`Flight RPC
+<./Flight>` framework.
+
+Generally, a database will implement the RPC methods according to the
+specification, but does not need to implement a client-side driver. A
+database client can use the provided Flight SQL client to interact
+with any database that supports the necessary endpoints.
```
Comment (jduo) [inline docs/source/format/FlightSql.rst:31]: Let's mention that Flight SQL clients decorate existing Flight client classes to add methods corresponding to the new RPC calls.

**Comment by lidavidm** [comment]
Thanks for the suggestions!

**Comment by pitrou** [comment]
> I would like to add sequence diagrams, how do we feel about something like [sphinxcontrib-mermaid](https://pypi.org/project/sphinxcontrib-mermaid/)?

Looks interesting. @jorisvandenbossche What do you think?

File: docs/source/format/FlightSql.rst [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+``ActionClosePreparedStatementRequest``
+    Close a previously created prepared statement.
+
+``ActionCreatePreparedStatementRequest``
+    Create a new prepared statement for a SQL query.
+
+``CommandPreparedStatementQuery``
+    Execute a previously created prepared statement and get the results.
+
+    When used with DoPut: binds parameter values to the prepared statement.
```
Comment (pitrou) [inline docs/source/format/FlightSql.rst:126]: So one must call first `DoPut` then `GetFlightInfo` to execute a query with parameters? Wouldn't it be nicer to be able to use `DoExchange` for that?

File: docs/source/format/FlightSql.rst [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+``type`` should be ``ClosePreparedStatement``).
+
+``ActionClosePreparedStatementRequest``
+    Close a previously created prepared statement.
+
+``ActionCreatePreparedStatementRequest``
+    Create a new prepared statement for a SQL query.
+
+``CommandPreparedStatementQuery``
+    Execute a previously created prepared statement and get the results.
```
Comment (pitrou) [inline docs/source/format/FlightSql.rst:124]: Does this also close the prepared statement or can it be reused multiple times?

**Review by pitrou — COMMENTED** [review]
Nice addition, thank you.

File: docs/source/format/FlightSql.rst (hunk 4 above)
Comment (lidavidm) [reply docs/source/format/FlightSql.rst:126]: I believe this was considered during the initial discussion but was rejected; @jduo do you remember why?

Comment (lidavidm) [reply docs/source/format/FlightSql.rst:126]: Ah, it was due to output parameters potentially changing the schema mid-stream. https://github.com/apache/arrow/pull/9368#discussion_r572941765

File: docs/source/format/FlightSql.rst (hunk 5 above)
Comment (lidavidm) [reply docs/source/format/FlightSql.rst:124]: It is reusable (else, you don't save much from having a prepared statement).

**Comment by lidavidm** [comment]
@jduo @pitrou any other comments here? I think we can use ARROW-14698 to expand the documentation once this lands

File: docs/source/format/FlightSql.rst [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+SQL Metadata
+------------
+
+Flight SQL provides a variety of commands to fetch catalog metadata
+about the database server.
+
+All of these commands can be used with the GetFlightInfo and GetSchema
+RPC methods. The Protobuf request message should be packed into a
+google.protobuf.Any message, then serialized and packed as the ``cmd``
+field in a CMD-type FlightDescriptor.
```
Comment (pitrou) [inline docs/source/format/FlightSql.rst:54]: For the record, would there be an easy way to cross-reference the table definitions in the Flight protobuf file (such as `FlightDescriptor` here)? I guess not, but just asking.

**Review by pitrou — APPROVED** [review]
+1

File: docs/source/format/FlightSql.rst (hunk 6 above)
Comment (lidavidm) [reply docs/source/format/FlightSql.rst:54]: Hmm, we would probably need to have Sphinx index the Protobuf file somehow and then use Sphinx's cross-reference support. Unfortunately there doesn't seem to be such integration from a quick glance (and if we wanted to do it, we would need to create a custom Protobuf plugin or modify the existing one (which supports neither Doxygen nor Sphinx): https://github.com/pseudomuto/protoc-gen-doc/

Comment (lidavidm) [reply docs/source/format/FlightSql.rst:54]: That may be worthwhile, though, since then we could avoid repeating docs between both files.

**Review by jduo — APPROVED** [review]

**Pull request closed without merging**